resources/net.go

```diff
@@ -0,0 +1,726 @@
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *NetRes) Default() Res {
+	return &NetRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+		State:   IfaceUp,
+		Timeout: time.Duration(10),
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *NetRes) Validate() error {
+	// validate state
+	if obj.State != IfaceUp && obj.State != IfaceDown {
+		return fmt.Errorf("invalid state: %s", obj.State)
+	}
+
+	// validate network address input
+	if obj.IPv4 != "" {
+		if _, _, err := net.ParseCIDR(obj.IPv4); err != nil {
+			return errwrap.Wrapf(err, "error parsing ipv4: %s", obj.IPv4)
+		}
+	}
+	if obj.IPv6 != "" {
+		if _, _, err := net.ParseCIDR(obj.IPv6); err != nil {
+			return errwrap.Wrapf(err, "error parsing ipv6: %s", obj.IPv6)
+		}
+	}
+	if obj.Gateway != "" {
+		if g := net.ParseIP(obj.Gateway); g == nil {
+			return fmt.Errorf("could not parse gateway: %s", obj.Gateway)
+		}
+	}
+
+	// validate the interface name
+	_, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+
+	return obj.BaseRes.Validate()
+}
+
+// Init runs some startup code for this resource.
+func (obj *NetRes) Init() error {
+	var err error
+	// get the network link's dbus path
+	obj.IfacePath, err = dbusIfacePath(obj.GetName())
+	if err != nil {
+		return errwrap.Wrapf(err, "error getting iface dbus path")
+	}
+
+	// establish a go-systemd dbus connection
+	obj.sdbus, err = systemd.New()
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to connect to systemd")
+	}
+	// establish a coreos godbus dbus connection
+	obj.gdbus, err = util.SystemBusPrivateUsable()
+	if err != nil {
+		return errwrap.Wrap(err, "error establishing system bus")
+	}
+
+	return obj.BaseRes.Init()
+}
+
+// Watch listens for networkd events by eavesdropping on the DBus Properties
+// interface (org.freedesktop.DBus.Properties) and filtering for signals from
+// the network device. It also watches the networkd service via a dbus unit
+// subscription, and the unit file responsible for configuring the interface.
+func (obj *NetRes) Watch() error {
+	var err error
+	// make sure systemd is running
+	if !systemdUtil.IsRunningSystemd() {
+		return fmt.Errorf("systemd is not running")
+	}
+
+	// create a dbus subscription to watch the networkd service unit
+	set := obj.sdbus.NewSubscriptionSet()
+	svcChan, svcErr := set.Subscribe() // svc event channel
+	set.Add(NetworkdService)
+
+	// Add a DBus rule to watch the "PropertiesChanged" signal and get messages
+	// from the network link's dbus path.
+	args := fmt.Sprintf(DBusEavesdropSprint, DBusPropertiesInterface, obj.IfacePath)
+	call := obj.gdbus.BusObject().Call(DBusAddMatch, 0, args)
+	if err := call.Err; err != nil {
+		return errwrap.Wrapf(err, "error creating dbus call")
+	}
+
+	// TODO: verify that implementation doesn't deadlock if there are unread
```

Comment:
This is a big TODO. If it's a serious thing, change this to a `XXX:` or a `FIXME:` if it's less serious.

Or we can just fix it :)